### PR TITLE
Add cost observability BQ table [VS-441]

### PR DESCRIPTION
**Pull request by mcovarr**

### CODE REVIEW

**Review by rsasch — APPROVED** [review]
LGTM 👍🏻

File: scripts/variantstore/wdl/GvsAssignIds.wdl [hunk 1]
```diff
@@ -19,7 +19,14 @@ workflow GvsAssignIds {
   String sample_info_table = "sample_info"
   String sample_info_schema_json = '[{"name": "sample_name","type": "STRING","mode": "REQUIRED"},{"name": "sample_id","type": "INTEGER","mode": "NULLABLE"},{"name":"is_loaded","type":"BOOLEAN","mode":"NULLABLE"},{"name":"is_control","type":"BOOLEAN","mode":"REQUIRED"},{"name":"withdrawn","type":"TIMESTAMP","mode":"NULLABLE"}]'
   String sample_load_status_json = '[{"name": "sample_id","type": "INTEGER","mode": "REQUIRED"},{"name":"status","type":"STRING","mode":"REQUIRED"}, {"name":"event_timestamp","type":"TIMESTAMP","mode":"REQUIRED"}]'
-
+  String cost_observability_json = '[ { "name": "call_set_identifier", "type": "STRING", "mode": "REQUIRED" }, ' +
```
Comment (RoriCremer) [inline scripts/variantstore/wdl/GvsAssignIds.wdl:22]: ok not for this pr---but I truly do think some of these would benefit from a description

**Review by RoriCremer — APPROVED** [review]

File: scripts/variantstore/wdl/GvsAssignIds.wdl [hunk 2]
```diff
@@ -45,6 +52,18 @@ workflow GvsAssignIds {
...
   }
 
+  call GvsCreateTables.CreateTables as CreateCostObservabilityTable {
+    input:
+      project_id = project_id,
+      dataset_name = dataset_name,
+      datatype = "cost_observability",
+      schema_json = cost_observability_json,
+      max_table_id = 1,
+      superpartitioned = "false",
```
[the remark this answers is not in the document]
Comment (mcovarr) [reply scripts/variantstore/wdl/GvsAssignIds.wdl:62]: or maybe on call set identifier?

File: scripts/variantstore/wdl/GvsAssignIds.wdl (hunk 1 above)
Comment (mcovarr) [reply scripts/variantstore/wdl/GvsAssignIds.wdl:22]: sure I can do that now

File: scripts/variantstore/wdl/GvsAssignIds.wdl (hunk 2 above)
[the remark this answers is not in the document]
Comment (RoriCremer) [reply scripts/variantstore/wdl/GvsAssignIds.wdl:62]: Yeah, that is great too

[the remark this answers is not in the document]
Comment (mcovarr) [reply scripts/variantstore/wdl/GvsAssignIds.wdl:62]: superpartitioning is the thing we do when we create a ref ranges or vet table for every 4000 samples to get around the BQ [partitioning limits](https://cloud.google.com/bigquery/quotas#partitioned_tables).

[the remark this answers is not in the document]
Comment (mcovarr) [reply scripts/variantstore/wdl/GvsAssignIds.wdl:62]: actually the `CreateTables` code is not flexible on the partition key so I'm going to make a larger revision here...

**Comment by mcovarr** [comment]
asking for a re-review since I ended up changing a lot...

**Review by rsasch — APPROVED** [review]
too bad you couldn't re-use the existing task to create the table, but LGTM 👍🏻

File: scripts/variantstore/wdl/GvsAssignIds.wdl [hunk 3]
```diff
@@ -168,3 +173,49 @@ task AssignIds {
...
+                                 '  { "name": "step", "type": "STRING", "mode": "REQUIRED" }, '                    + # The name of the core GVS workflow to which this belongs
+                                 '  { "name": "call", "type": "STRING", "mode": "NULLABLE" }, '                    + # The WDL call to which this belongs
+                                 '  { "name": "shard_identifier", "type": "STRING", "mode": "NULLABLE" }, '        + # A unique identifier for this shard, may or may not be its index
+                                 '  { "name": "call_start_timestamp", "type": "TIMESTAMP", "mode": "REQUIRED" }, ' + # When the call logging this event was started
+                                 '  { "name": "event_timestamp", "type": "TIMESTAMP", "mode": "REQUIRED" }, '      + # When the observability event was logged
+                                 '  { "name": "event_key", "type": "STRING", "mode": "REQUIRED" }, '               + # The type of observability event being logged
+                                 '  { "name": "event_bytes", "type": "INTEGER", "mode": "REQUIRED" } ] '             # Number of bytes reported for this observability event
```
Comment (RoriCremer) [inline scripts/variantstore/wdl/GvsAssignIds.wdl:190]: hahaha I meant that the "description" could go in the BQ table, but this works perfectly too

File: scripts/variantstore/wdl/GvsAssignIds.wdl [hunk 4]
```diff
@@ -168,3 +173,49 @@ task AssignIds {
...
+
+    if [ $BQ_SHOW_RC -ne 0 ]; then
+      PARTITION_STRING="--time_partitioning_field call_start_timestamp --time_partitioning_type DAY"
+      echo "making table $TABLE"
+      echo '~{cost_observability_json}' > schema.json
+      bq --location=US mk ${PARTITION_STRING} --project_id=~{project_id} $TABLE schema.json
+    fi
+  >>>
+  runtime {
+    docker: "us.gcr.io/broad-dsde-methods/variantstore:ah_var_store_2022_05_16"
```
Comment (RoriCremer) [inline scripts/variantstore/wdl/GvsAssignIds.wdl:215]: why are we using our custom docker?

I'm worried about this getting stale (though I suppose it wont matter for this use case)

I feel like we tend to use:     docker: "us.gcr.io/broad-gatk/gatk:4.1.7.0"

File: scripts/variantstore/wdl/GvsAssignIds.wdl (hunk 3 above)
Comment (RoriCremer) [inline scripts/variantstore/wdl/GvsAssignIds.wdl:190]:
```suggestion
  String cost_observability_json =
                                 '[ { "name": "call_set_identifier", "type": "STRING", "mode": "REQUIRED", "description": "The name by which we refer to the callset" }, '     +
                                 '  { "name": "step", "type": "STRING", "mode": "REQUIRED", "description": "The name of the core GVS workflow to which this belongs" }, '                    +
                                 '  { "name": "call", "type": "STRING", "mode": "NULLABLE", "description": "The WDL call to which this belongs"  }, '                    +
                                 '  { "name": "shard_identifier", "type": "STRING", "mode": "NULLABLE" , "description": "A unique identifier for this shard, may or may not be its index" }, '        +
                                 '  { "name": "call_start_timestamp", "type": "TIMESTAMP", "mode": "REQUIRED", "description": "When the call logging this event was started" }, ' +
                                 '  { "name": "event_timestamp", "type": "TIMESTAMP", "mode": "REQUIRED", "description": "When the observability event was logged"}, '      +
                                 '  { "name": "event_key", "type": "STRING", "mode": "REQUIRED", "description": "The type of observability event being logged" }, '               +
                                 '  { "name": "event_bytes", "type": "INTEGER", "mode": "REQUIRED", "description": "Number of bytes reported for this observability event" } ] '

**Pull request merged**